plutus-core/plutus-ir/test/GeneratorSpec/Types.hs

```diff
@@ -0,0 +1,65 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+module GeneratorSpec.Types where
+
+import Data.Either
+import Data.Map qualified as Map
+import PlutusCore.Generators.PIR
+import Test.QuickCheck
+
+-- | Check that the types we generate are kind-correct
+-- See Note [Debugging generators that don't generate well-typed/kinded terms/types]
+-- and see the utility tests below when this property fails.
+prop_genKindCorrect :: Bool -> Property
+prop_genKindCorrect debug =
+  -- Context minimality doesn't help readability, so no shrinking here
+  forAllDoc "ctx" genCtx (const []) $ \ ctx ->
+  -- Note, no shrinking here because shrinking relies on well-kindedness.
+  forAllDoc "k,ty" (if debug then genKindAndTypeDebug else genKindAndType) (const []) $ \ (k, ty) ->
+  checkKind ctx ty k
+
+-- | Check that shrinking types maintains kinds
+prop_shrinkTypeSound :: Property
+prop_shrinkTypeSound =
+  forAllDoc "ctx" genCtx (const []) $ \ ctx ->
+  forAllDoc "k,ty" (genKindAndTypeWithCtx ctx) (shrinkKindAndType ctx) $ \ (k, ty) ->
+  -- See discussion about the same trick in `prop_shrinkTermSound`
+  isRight (checkKind ctx ty k) ==>
+  assertNoCounterexamples $ lefts
+    [ (k, ty, ) <$> checkKind ctx ty k
+    | (k, ty) <- shrinkKindAndType ctx (k, ty)
+    ]
+
+-- Utility tests for debugging
+
+-- | Test that shrinking types results in smaller types. Useful for debugging shrinking.
+prop_shrinkTypeSmallerKind :: Property
```

Comment:
Yes, probably. It says that when you shrink a type you end up with a type at a smaller kind. We don't track shrinking order on types because writing it down is *hard work* (also type shrinking order is probably undecidable anywa) so we can't test it.